src/runtime/GlobalObjectBuiltinObject.cpp

```diff
@@ -90,65 +89,70 @@ static Value builtinObjectPreventExtensions(ExecutionState& state, Value thisVal
-        } else {
-            StringBuilder builder;
-            builder.appendString("[object ");
-            builder.appendString(tag);
-            builder.appendString("]");
-            return AtomicString(state, builder.finalize()).string();
-        }
+    Value builtinTag;
+    if (thisObject->isArray(state)) {
+        builtinTag = strings->lazyObjectArrayToString().string();
+    } else if (thisObject->isArgumentsObject()) {
+        builtinTag = strings->lazyObjectArgumentsToString().string();
+    } else if (thisObject->isCallable()) {
+        builtinTag = strings->lazyObjectFunctionToString().string();
+    } else if (thisObject->isErrorObject()) {
+        builtinTag = strings->lazyObjectErrorToString().string();
+    } else if (thisObject->isBooleanObject()) {
+        builtinTag = strings->lazyObjectBooleanToString().string();
+    } else if (thisObject->isNumberObject()) {
+        builtinTag = strings->lazyObjectNumberToString().string();
+    } else if (thisObject->isStringObject()) {
+        builtinTag = strings->lazyObjectStringToString().string();
+    } else if (thisObject->isDateObject()) {
+        builtinTag = strings->lazyObjectDateToString().string();
+    } else if (thisObject->isRegExpObject()) {
+        builtinTag = strings->lazyObjectRegExpToString().string();
     } else {
-        if (thisObject->isArray(state)) {
-            return state.context()->staticStrings().lazyObjectArrayToString().string();
-        } else if (thisObject->isStringObject()) {
-            return state.context()->staticStrings().lazyObjectStringToString().string();
-        } else if (thisObject->isArgumentsObject()) {
-            return state.context()->staticStrings().lazyObjectArgumentsToString().string();
-        } else if (thisObject->isCallable()) {
-            return state.context()->staticStrings().lazyObjectFunctionToString().string();
-        } else if (thisObject->isErrorObject()) {
-            return state.context()->staticStrings().lazyObjectErrorToString().string();
-        } else if (thisObject->isBooleanObject()) {
-            return state.context()->staticStrings().lazyObjectBooleanToString().string();
-        } else if (thisObject->isNumberObject()) {
-            return state.context()->staticStrings().lazyObjectNumberToString().string();
-        } else if (thisObject->isDateObject()) {
-            return state.context()->staticStrings().lazyObjectDateToString().string();
-        } else if (thisObject->isRegExpObject()) {
-            return state.context()->staticStrings().lazyObjectRegExpToString().string();
-        } else {
-            return state.context()->staticStrings().lazyObjectObjectToString().string();
-        }
+        builtinTag = strings->lazyObjectObjectToString().string();
     }
+
+    Value toStringTag = thisObject->get(state, ObjectPropertyName(state.context()->vmInstance()->globalSymbols().toStringTag)).value(state, thisObject);
+    if (!toStringTag.isString()) {
```

Comment:
How about read built-in tag in here?
IMO multiple type checking is slow

Reply:
Good idea!
I'll change it as you commented, but `isArray` check should be run first because of `Proxy` case.
Other check code seems fine to run later.